src/index.rs

```diff
@@ -0,0 +1,193 @@
+                        // bump of cargo, but that's unlikely (famous last words)
+                        let minor = semver.split('.').nth(1).ok_or_else(|| {
+                            io::Error::new(
+                                io::ErrorKind::InvalidData,
+                                "cargo semver was in an invalid format",
+                            )
+                        })?;
+
+                        let minor: u32 = minor
+                            .parse()
+                            .map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))?;
+
+                        minor >= 70
+                    }
+                }
+            }
+        };
+
+        let url = if sparse_index {
+            crate::CRATES_IO_HTTP_INDEX
+        } else {
+            crate::INDEX_GIT_URL
+        };
+
+        let (path, canonical) = dirs::get_index_details(url, cargo_home)?;
+
+        if sparse_index {
+            Ok(Self::Sparse(SparseIndex::at_path(path, canonical)))
+        } else {
+            Index::with_path(path, canonical).map(Self::Git)
+        }
+    }
+
+    /// Retrieves the current version of cargo being used
+    pub fn cargo_version() -> Result<String, Error> {
+        let mut cargo = std::process::Command::new(
+            std::env::var_os("CARGO")
+                .as_deref()
+                .unwrap_or_else(|| std::ffi::OsStr::new("cargo")),
+        );
+
+        cargo.arg("-V");
+        cargo.stdout(std::process::Stdio::piped());
+
+        let output = cargo.output()?;
+        if !output.status.success() {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                "failed to request cargo version information",
+            )
+            .into());
+        }
+
+        let stdout = String::from_utf8(output.stdout)
+            .map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))?;
+
+        let semver = stdout.split(' ').nth(1).ok_or_else(|| {
+            io::Error::new(
+                io::ErrorKind::InvalidData,
+                "cargo version information was in an invalid format",
+            )
+        })?;
+
+        Ok(semver.to_owned())
+    }
+
+    /// Retrieves the index metadata for the specified crate name
+    #[inline]
+    pub fn krate(&self, name: &str) -> Result<crate::Crate, Error> {
+        match self {
+            Self::Git(index) => index.crate_(name).ok_or_else(|| {
+                io::Error::new(
+                    io::ErrorKind::NotFound,
+                    format!("unable to locate crate '{name}'"),
+                )
+                .into()
+            }),
+            Self::Sparse(index) => index.crate_from_cache(name),
+        }
+    }
+
+    /// If using the sparse index, returns the url that can be used to fetch
+    /// the index entry for the specified crate
+    #[inline]
+    pub fn crate_url(&self, name: &str) -> Option<String> {
```

Comment:
because it will always fail with one index type, this is not an abstraction, only a layer of indirection. IMHO that's not a good API. Users who need the URLs should be working with sparseindex directly.